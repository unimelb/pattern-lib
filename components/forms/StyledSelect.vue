<template>
  <div :id="null">
    <select
      :id="id"
      v-model="content"
      aria-label="choose an option"
      :class="extraclasses"
      @input="handleInput"
      @change="callback">
      <slot />
      <option
        v-for="(option, index) in optionsList"
        :key="index"
        :value="option.value">
        {{ option.label }}
      </option>
    </select>
    <div
      v-if="label"
      class="styled-select__label">
      {{ label }}
    </div>
  </div>
</template>

<script>
import isPlainObject from 'lodash.isplainobject';

export default {
  name: 'StyledSelect',
  props: {
    id: {
      type: [String, Boolean],
      default: () => false,
    },
    value: {
      type: String,
      default: '',
    },
    blue: {
      type: Boolean,
      default: false,
    },
    label: {
      type: String,
      default: '',
    },
    options: {
      type: Array,
      default: () => [],
    },
    callback: {
      type: Function,
      default: () => {},
    },
  },
  data: () => ({
    extraclasses: '',
    baseclass: 'styled-select',
    blueclass: 'styled-select--blue',
    content: '',
  }),
  computed: {
    optionsList() {
      const { options } = this;

      // If options are not an object then convert them to label/value object
      const newOptions = options.map((option) => {
        if (!isPlainObject(option)) {
          return {
            label: option,
            value: option,
          };
        }

        return option;
      });

      return newOptions;
    },
  },
  watch: {
    options(newOptions) {
      const [defaultOption] = newOptions;

      this.content = defaultOption;
    },
    value(newOptions) {
      this.content = newOptions;
    },
  },
  mounted() {
    this.extraclasses = this.$el.className;
    this.$el.className = `${this.baseclass}${this.blue ? ` ${this.blueclass}` : ''}`;

    this.content = this.setDefaultValue();
  },
  methods: {
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
    setDefaultValue() {
      const defaultSlot = this.$slots.default;
      const filterResult = (items) => items.filter((child) => child.tag === 'option');
      const filterDefaultValue = (items) => items.map((value) => value.children[0].text);

      if (defaultSlot && defaultSlot.length) {
        // Only grab <option>
        const children = filterResult(defaultSlot);

        // Grab value of first item.
        const [defaultValue] = filterDefaultValue(children);

        return defaultValue.trim();
      }

      return '';
    },
  },
};
</script>
