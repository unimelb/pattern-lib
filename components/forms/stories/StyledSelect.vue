<template>
  <form class="form">
    <h2>Loose</h2>
    <div class="row">
      <label for="f_q0">Type</label>
      <StyledSelect id="f_q0">
        <option>Undergraduate</option>
        <option>Postgraduate</option>
        <option>Reasearch</option>
        <option>Non-award</option>
      </StyledSelect>
    </div>
    <div class="row">
      <label for="f_q1">Type</label>
      <StyledSelect
        id="f_q1"
        blue>
        <option>Undergraduate</option>
        <option>Postgraduate</option>
        <option>Reasearch</option>
        <option>Non-award</option>
      </StyledSelect>
    </div>

    <h2>With label</h2>
    <div class="row">
      <label for="f_q0">Type</label>
      <StyledSelect
        id="f_q0"
        label="Lorem Ipsum">
        <option>Undergraduate</option>
        <option>Postgraduate</option>
        <option>Reasearch</option>
        <option>Non-award</option>
      </StyledSelect>
    </div>

    <SectionWrap class="bg-white">
      <h2>Enclosed in a section</h2>
      <div class="row">
        <label for="f_q2">Type</label>
        <StyledSelect
          id="f_q2"
          blue>
          <option>Undergraduate</option>
          <option>Postgraduate</option>
          <option>Reasearch</option>
          <option>Non-award</option>
        </StyledSelect>
      </div>

      <h2 class="shim-pt2 shim-mb05">
        Grid
      </h2>
      <div class="grid">
        <div class="cell cell--tab-1of3">
          <label for="f_q3">Type</label>
          <StyledSelect
            id="f_q3"
            blue>
            <option>Undergraduate</option>
            <option>Postgraduate</option>
            <option>Reasearch</option>
            <option>Non-award</option>
          </StyledSelect>
        </div>
        <div class="cell cell--tab-2of3">
          <label for="f_q4">Type</label>
          <StyledSelect
            id="f_q4"
            blue>
            <option>Undergraduate</option>
            <option>Postgraduate</option>
            <option>Reasearch</option>
            <option>Non-award</option>
          </StyledSelect>
        </div>
      </div>
    </SectionWrap>
  </form>
</template>

<script>
import StyledSelect from '../StyledSelect.vue';

export default {
  components: { StyledSelect },
  readme: { html: true, source: false },
};
</script>
