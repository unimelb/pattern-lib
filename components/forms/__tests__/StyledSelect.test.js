import {
  shallow,
} from 'vue-test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import StyledSelect from '../StyledSelect.vue';

expect.extend(toHaveNoViolations);

describe('StyledSelect', () => {
  it('should match snapshot', () => {
    const result = shallow(StyledSelect, {
      propsData: {
        label: 'text',
      },
    }).element;

    expect(result).toMatchSnapshot();
  });

  it('should show label', () => {
    const wrapper = shallow(StyledSelect, {
      propsData: {
        label: 'text',
      },
    });

    expect(wrapper.find('.styled-select__label').text()).toBe('text');
    expect(wrapper.find('.styled-select__label').exists()).toBe(true);
  });

  it('Component throws no accessibility violations', (done) => {
    const html = shallow(StyledSelect).html();
    // pass anything that outputs html to axe
    return axe(html).then((response) => {
      expect(response).toHaveNoViolations();
      done();
    });
  });
});
