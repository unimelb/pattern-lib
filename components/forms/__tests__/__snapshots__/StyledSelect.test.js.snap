// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledSelect should match snapshot 1`] = `
<div
  class="styled-select"
>
  <select
    aria-label="choose an option"
    class=""
  >
     
  </select>
   
  <div
    class="styled-select__label"
  >
    
    text
  
  </div>
</div>
`;
